kura/org.eclipse.kura.core.configuration/src/main/java/org/eclipse/kura/core/configuration/ConfigurationServiceImpl.java

```diff
@@ -1006,28 +1009,42 @@ private void writeSnapshot(long sid, XmlComponentConfigurations conf) throws Kur
             throw new KuraException(KuraErrorCode.CONFIGURATION_SNAPSHOT_NOT_FOUND);
         }
 
-        // Marshall the configuration into an XML
-        String xmlResult;
-        xmlResult = marshal(conf);
-        if (xmlResult == null || xmlResult.trim().isEmpty()) {
-            throw new KuraException(KuraErrorCode.INVALID_PARAMETER, conf);
+        if (fSnapshot.isDirectory()) {
+            throw new KuraException(KuraErrorCode.CONFIGURATION_ERROR, fSnapshot.getName() + "is not a file.");
+        }
+
+        File tempSnapshotFile;
+        try {
+            tempSnapshotFile = File.createTempFile(fSnapshot.getName(), "", new File(fSnapshot.getParent()));
+            tempSnapshotFile.deleteOnExit();
+        } catch (IOException ex) {
+            throw new KuraIOException(ex);
         }
 
-        // Encrypt the XML
-        char[] encryptedXML = this.cryptoService.encryptAes(xmlResult.toCharArray());
+        // Write the temporary snapshot
+        try (FileOutputStream fos = new FileOutputStream(tempSnapshotFile);
+                OutputStream encryptedStream = this.cryptoService.aesEncryptingStream(fos)) {
 
-        // Write the snapshot
-        try (FileOutputStream fos = new FileOutputStream(fSnapshot);
-                OutputStreamWriter osw = new OutputStreamWriter(fos, StandardCharsets.UTF_8);) {
             logger.info("Writing snapshot - Saving {}...", fSnapshot.getAbsolutePath());
-            osw.append(new String(encryptedXML));
-            osw.flush();
+
+            marshal(encryptedStream, conf);
+            encryptedStream.flush();
             fos.flush();
             fos.getFD().sync();
+
             logger.info("Writing snapshot - Saving {}... Done.", fSnapshot.getAbsolutePath());
         } catch (IOException e) {
             throw new KuraIOException(e);
         }
+
+        try {
+            // Consolidate snapshot writing
+            FileUtils.copyFile(tempSnapshotFile, fSnapshot);
```

Comment:
Please use Files.move with the StandardCopyOptions.REPLACE_EXISTING and StandardCopyOptions.ATOMIC_MOVE flags to do this